### ISSUE TITLE
Specifying a distribution with moments?

### ISSUE THREAD

**Issue by ahwillia** [JuliaStats/Distributions.jl#443]
I'm writing a package now where it would be convienent/natural to specify a `NegativeBinomial` with the mean and variance rather than `r` and `p`. I wrote a wrapper function like this:

```julia
function fit_NegBinomial(mu,sigma2)
	p = mu / sigma2
	r = mu*p / (1-p)
	return NegativeBinomial(r,p)
end
```

While this works, I find it a little clunky. My gut says it would be preferable to write a new constructor:

```julia
function NegativeBinomial(mu::Real,sigma2::Real)
    ...
end
```

The problem of course is that there would be no way to distinguish between this and the existing constructor that takes `r` and `p` as inputs... :disappointed:. Maybe one way to do this would be to add a flag telling the constructor how to interpret the inputs

```julia
function NegativeBinomial(a, b; params=:moments)
    ...
end
```

That is when `params == :moments` the constructor knows to interpret `a` and `b` as the mean and variance.

Or maybe I'm just fretting over small matters.

**Comment by simonbyrne** [comment]
This is something we probably should have (it's typically called method of moments, or moment matching).

**Comment by johnmyleswhite** [comment]
I don't think Alex is referring to the method moments because there's no data. I think he just wants to specify a distribution like the `Gamma` distribution using its mean and variance rather than using its shape and scale.

I have some code for doing this for a few distributions. I've called the function `meanvar` in the past.

One example:

```
function meanvar(::Type{LogNormal}, m, v)
    z = v / m^2 + 1
    LogNormal(
        log(m) - 0.5 * log(z),
        sqrt(log(z)),
    )
end
```

**Comment by kleinschmidt** [comment]
Defining something like a `Moments` type and then defining `convert` methods might also work. And if we make Moments an abstract type (or parametrize by the number of moments) that could also support nice error/warning messages when you ask, for example, for a Poisson distribution with an incompatible mean and variance...

**Comment by ahwillia** [comment]
I kind of like the simplicity of the `meanvar` method. Although the name might need to be changed to be more general (e.g., a Poisson distribution wouldn't have the variance specified).

@kleinschmidt -- maybe you could post a short example of what you are proposing exactly. Would it be along the lines of:

```julia
m = Moments(1.0,0.5) # specify mean = 1.0, variance = 0.5
d = NegativeBinomial(m)
```

**Comment by kleinschmidt** [comment]
Now that I've thought a bit more about it, I think I also prefer the `meanvar` method. Because it's dispatched on the distribution, you can have a different signature for each distribution as needed.

**Comment by ahwillia** [comment]
Maybe we could call it:

* `match_moments(...)`
* `fit_moments(...)`
* `specify_moments(...)`?

Or maybe just:

* `match(...)`
* `specify(...)`?

I'm happy to open a pull request that implements this for a few common distributions.

**Comment by johnmyleswhite** [comment]
I think we should try to understand the full space of parameterizations we want to express before choosing an API. Specifying a distribution by moments is useful, but perhaps we want other specifications as well (e.g. natural parameters of an exponential family).

**Comment by ahwillia** [comment]
Good point. But would this involve separate functions for each parameterization choice? If we want to support natural parameters, moments, and potentially others forms we could do something like:

```julia
specify(::Distribution,::Symbol, args...) = ...

# example for moments
specify(NegativeBinomial, :moments, m, v)

# example for natural params
specify(Poisson, :natural, n)
```

Are there more exotic parameterizations that wouldn't fit into this scheme? My stats knowledge is maybe limited on this.

**Comment by simonbyrne** [comment]
I like `frommeanvar` (I would expect `meanvar` to give the tuple `(mean(d), var(d))`).

**Comment by andreasnoack** [comment]
I like overloading the constructed with methods for a `Moments` type better but I must also admit that I haven't thought this through yet.
